carla/recourse_methods/catalog/cem/cem.py

```diff
@@ -0,0 +1,578 @@
+        instances = factuals.drop(columns=[target_name])
+
+        # normalize and one-hot-encoding
+        instances = self.catalog_model.perform_pipeline(instances)
+        instances = instances[self.catalog_model.feature_input_order]
+
+        counterfactuals = []
+        times_list = []
+
+        for i in range(instances.values.shape[0]):
+            start = timeit.default_timer()
+            _, counterfactual = self.counterfactual_search(instances.values[i, :])
+            stop = timeit.default_timer()
+            time_taken = stop - start
+
+            counterfactuals.append(counterfactual)
+            times_list.append(time_taken)
+
+        counterfactuals_df = pd.DataFrame(np.array(counterfactuals))
+        counterfactuals_df.columns = instances.columns
+
+        # Success rate & drop not successful counterfactuals & process remainder
+        success_rate, counterfactuals_indices = success_rate_and_indices(
+            counterfactuals_df
+        )
+        counterfactuals_df = counterfactuals_df.iloc[counterfactuals_indices]
+        instances = instances.iloc[counterfactuals_indices]
+
+        # Obtain labels
+        instance_label = np.argmax(self.catalog_model.predict(instances.values), axis=1)
+        counterfactual_label = np.argmax(
+            self.catalog_model.predict(counterfactuals_df.values), axis=1
+        )
+
+        # Order counterfactuals and instances in original data order
+        counterfactuals_df = counterfactuals_df[self.catalog_model.feature_input_order]
+        instances = instances[self.catalog_model.feature_input_order]
+
+        categorical_cols = self.data.categoricals
+        if len(categorical_cols) > 0:
+            # Convert binary cols of counterfactuals and instances into strings: Required for >>Measurement<< in script
+            # Convert binary cols back to original string encoding
+            fitted_encoder = self.catalog_model.encoder
+            counterfactuals_df = decode(
+                fitted_encoder, categorical_cols, counterfactuals_df
+            )
+            instances = decode(fitted_encoder, categorical_cols, instances)
+
+        # Add labels
+        counterfactuals_df[target_name] = counterfactual_label
+        instances[target_name] = instance_label
+
+        # Collect in list making use of pandas
+        instances_list = []
+        counterfactuals_list = []
+
+        for i in range(counterfactuals_df.shape[0]):
+            counterfactuals_list.append(
+                pd.DataFrame(
+                    counterfactuals_df.iloc[i].values.reshape((1, -1)),
+                    columns=counterfactuals_df.columns,
+                )
+            )
+            instances_list.append(
+                pd.DataFrame(
+                    instances.iloc[i].values.reshape((1, -1)), columns=instances.columns
+                )
+            )
+
+        return instances_list, counterfactuals_list, times_list, success_rate
```

Comment:
At the moment our output for `get_counterfactual()` would be only the counterfactual encoded and normalized dataframe. Regarding the measurement outputs time and success rate, it would be nice to compute them only by calling the benchmarking methods.

Reply:
fixed. Do you know why the dataframes are converted to lists here? Is this important for something else, or can I just delete this code?

Reply:
This is some remnant from the last project. Now we need the output counterfactuals as DataFrames, and it is important that every counterfactual flips the prediction of the black box model. If not, the row is replaced with an empty one.